File: src/engine/search/repository.py
"""
KIVO Search Engine
Entry Repository

Reine Persistenz. Kein Ranking, kein Scoring, keine Business-Logik.
'search' hier ist ein einfacher Rohzugriff (Substring-Filter), NICHT
die vom Client aufgerufene, gerankte Suche - die laeuft immer ueber
die Pipeline in engine.py.
"""

from __future__ import annotations

from uuid import UUID

from .entry import Entry


class EntryRepository:

    def __init__(self) -> None:
        self._entries: dict[UUID, Entry] = {}

    def create(self, entry: Entry) -> Entry:
        self._entries[entry.id] = entry
        return entry

    def update(self, entry: Entry) -> Entry:
        entry.touch()
        self._entries[entry.id] = entry
        return entry

    def delete(self, entry_id: UUID) -> None:
        self._entries.pop(entry_id, None)

    def get(self, entry_id: UUID) -> Entry | None:
        return self._entries.get(entry_id)

    def get_all(self) -> tuple[Entry, ...]:
        return tuple(self._entries.values())

    def search(self, term: str) -> tuple[Entry, ...]:
        """Roh-Substring-Suche ueber Titel + Inhalt (case-insensitive)."""
        term = term.lower()
        return tuple(
            e for e in self._entries.values()
            if term in e.title.lower() or term in e.content.lower()
        )

File: src-tauri/src/engine/repository.rs
//! KIVO Search Engine (Rust-Port)
//! Entry Repository
//!
//! 1:1-Port von `src/engine/search/repository.py`.
//! Reine In-Memory-Verwaltung. Kein Ranking, kein Scoring, keine
//! Business-Logik.
//!
//! KORREKTUR (Phase 3, siehe context.rs): nutzt jetzt `IndexMap` statt
//! `HashMap`, damit `get_all()` die Entries in Einfuegereihenfolge liefert -
//! exakt wie Python's insertion-ordered Dict. Das ist Voraussetzung dafuer,
//! dass SearchStage/ResultBuilder bei Score-Gleichstand dieselbe
//! Tie-Break-Reihenfolge liefern wie das Original.

use indexmap::IndexMap;
use uuid::Uuid;

use crate::engine::entry::Entry;

pub struct EntryRepository {
    entries: IndexMap<Uuid, Entry>,
}

impl EntryRepository {
    pub fn new() -> Self {
        Self {
            entries: IndexMap::new(),
        }
    }

    pub fn create(&mut self, entry: Entry) -> Entry {
        let clone = entry.clone();
        self.entries.insert(entry.id, entry);
        clone
    }

    pub fn update(&mut self, mut entry: Entry) -> Entry {
        entry.touch();
        let clone = entry.clone();
        self.entries.insert(entry.id, entry);
        clone
    }

    pub fn delete(&mut self, entry_id: Uuid) {
        self.entries.shift_remove(&entry_id);
    }

    pub fn get(&self, entry_id: Uuid) -> Option<Entry> {
        self.entries.get(&entry_id).cloned()
    }

    pub fn get_all(&self) -> Vec<Entry> {
        self.entries.values().cloned().collect()
    }

    /// Entspricht der rohen Substring-Suche in repository.py. Wird von der
    /// gerankten Pipeline NICHT verwendet, nur der Vollstaendigkeit halber
    /// mit portiert.
    #[allow(dead_code)]
    pub fn search(&self, term: &str) -> Vec<Entry> {
        let term = term.to_lowercase();
        self.entries
            .values()
            .filter(|e| {
                e.title.to_lowercase().contains(&term) || e.content.to_lowercase().contains(&term)
            })
            .cloned()
            .collect()
    }
}

impl Default for EntryRepository {
    fn default() -> Self {
        Self::new()
    }
}
